cmd/trtlsim/main.go

```diff
@@ -0,0 +1,238 @@
+	chunkSize = 512                                                    // size of each write
+)
+
+// dummy namespaces for simulation
+var namespaces = []string{"catchfireBarons", "falselightCutters", "fullCrowns", ""}
+
+// probabilities for reads/writes/deletes
+var probabilities = map[string]uint{
+	"read":   60,
+	"write":  38,
+	"delete": 2,
+}
+
+func main() {
+
+	// assumes trtl is already being served (e.g. from the trtl cli)
+	// sim needs the endpoint (e.g. localhost:port) + certs (just stubs for now)
+	// run trtl in insecure mode
+	sim := new(endpoint, insecure)
+	simClient, err := sim.connect()
+	if err != nil {
+		panic(err)
+	}
+
+	ticker := jitter.New(interval, sigma)
+
+	// start on the first tick
+	for ; true; <-ticker.C {
+
+		// multiple regions access the data store concurrently; 1 routine per region
+		for r := 1; r <= regions; r++ {
+			go sim.accessor(simClient)
+		}
+	}
+}
+
+type Simulator struct {
+	Endpoint string      `yaml:"endpoint"`           // the replica endpoint to connect to
+	Insecure bool        `yaml:"insecure,omitempty"` // do not connect with TLS
+	Selector *wr.Chooser `yaml:"chooser,omitempty"`  // random selection helper
+	CertPath string      `yaml:cert_path,omitempty"` // path to certificate key pair for client side mTLS
+	PoolPath string      `yaml:poo_path,omitempty"`  // path to certificate trust chain for client side mTLS
+}
+
+func new(endpoint string, insecure bool) *Simulator {
+	// initialize weighted probability selector
+	selector := initialize()
+	return &Simulator{
+		Endpoint: endpoint,
+		Insecure: insecure,
+		Selector: selector,
+		CertPath: certPath,
+		PoolPath: poolPath,
+	}
+}
+
+// Connect to the trtl server and return a gRPC client
+func (s *Simulator) connect() (_ pb.TrtlClient, err error) {
+	var opts []grpc.DialOption
+	if s.Insecure {
+		opts = append(opts, grpc.WithInsecure())
+	} else {
+		var sz *trust.Serializer
+		if sz, err = trust.NewSerializer(false); err != nil {
+			return nil, err
+		}
+
+		var pool trust.ProviderPool
+		if pool, err = sz.ReadPoolFile(s.PoolPath); err != nil {
+			return nil, err
+		}
+
+		var provider *trust.Provider
+		if provider, err = sz.ReadFile(s.CertPath); err != nil {
+			return nil, err
+		}
+
+		var cert tls.Certificate
+		if cert, err = provider.GetKeyPair(); err != nil {
+			return nil, err
+		}
+
+		var certPool *x509.CertPool
+		if certPool, err = pool.GetCertPool(false); err != nil {
+			return nil, err
+		}
+		var u *url.URL
+		if u, err = url.Parse(s.Endpoint); err != nil {
+			return nil, err
+		}
+		conf := &tls.Config{
+			ServerName:   u.Host,
+			Certificates: []tls.Certificate{cert},
+			RootCAs:      certPool,
+		}
+		opts = append(opts, grpc.WithTransportCredentials(credentials.NewTLS(conf)))
+	}
+
+	// Connect the replica client
+	var cc *grpc.ClientConn
+	if cc, err = grpc.Dial(s.Endpoint, opts...); err != nil {
+		return nil, err
+	}
+	return pb.NewTrtlClient(cc), nil
```

Comment:
To figure out what problems we might be having in the deployment, could we add a log statement before we return the client, e.g. something like `log.Printf("connected to trtl server at %s\n", s.Endpoint)`. If we've gotten the configuration wrong, this will help us quickly understand what's going on from the k8s output.